xla/service/gpu/ir_emitter_unnested.cc

```diff
@@ -729,78 +753,135 @@ absl::Status IrEmitterUnnested::EmitCublasLtMatmulThunk(
     const HloCustomCallInstruction* instr) {
-  TF_RET_CHECK(instr->operand_count() == 6 || instr->operand_count() == 7 ||
-               instr->operand_count() == 8);
   TF_ASSIGN_OR_RETURN(const auto gpu_config,
                       instr->backend_config<xla::gpu::GpuBackendConfig>());
   xla::gpu::GemmBackendConfig config = gpu_config.gemm_backend_config();
   xla::gpu::GemmBackendConfig_Epilogue epilogue = config.epilogue();
+  TF_RET_CHECK(instr->operand_count() == 6 || instr->operand_count() == 7 ||
+               instr->operand_count() == 8);
 
+  bool has_tuple_output = instr->shape().IsTuple();
+  TF_ASSIGN_OR_RETURN(
+      bool has_aux_output,
+      xla::gpu::gpublas_lt::EpilogueHasAuxiliaryOutput(epilogue));
+  TF_ASSIGN_OR_RETURN(
+      bool has_aux_input,
+      xla::gpu::gpublas_lt::EpilogueHasAuxiliaryInput(epilogue));
   TF_ASSIGN_OR_RETURN(bool has_vector_bias,
                       xla::gpu::gpublas_lt::EpilogueAddsVectorBias(epilogue));
-  bool has_damax = instr->shape().IsTuple();
+  bool has_matrix_bias = config.beta() != 0;
+  bool is_relu_epilogue = (epilogue == GemmBackendConfig::RELU_AUX ||
+                           epilogue == GemmBackendConfig::D_RELU ||
+                           epilogue == GemmBackendConfig::BIAS_RELU_AUX ||
+                           epilogue == GemmBackendConfig::D_RELU_BGRAD);
   xla::ShapeIndex output_index =
-      has_damax ? xla::ShapeIndex{0} : xla::ShapeIndex{};
-
+      has_tuple_output ? xla::ShapeIndex{0} : xla::ShapeIndex{};
   TF_ASSIGN_OR_RETURN(BufferAllocation::Slice a,
                       GetAllocationSliceForHlo(instr->operand(0)));
   TF_ASSIGN_OR_RETURN(BufferAllocation::Slice b,
                       GetAllocationSliceForHlo(instr->operand(1)));
   BufferAllocation::Slice c;
-  bool has_matrix_bias = config.beta() != 0;
   if (has_matrix_bias) {
     TF_ASSIGN_OR_RETURN(c, GetAllocationSliceForHlo(instr->operand(2)));
   } else {
     TF_ASSIGN_OR_RETURN(c, GetAllocationSliceForHlo(instr, output_index));
   }
+  int a_scale_index = has_matrix_bias ? 3 : 2;
   TF_ASSIGN_OR_RETURN(BufferAllocation::Slice d,
                       GetAllocationSliceForHlo(instr, output_index));
-
-  int a_scale_index = has_matrix_bias ? 3 : 2;
   TF_ASSIGN_OR_RETURN(BufferAllocation::Slice a_scale,
                       GetAllocationSliceForHlo(instr->operand(a_scale_index)));
   TF_ASSIGN_OR_RETURN(
       BufferAllocation::Slice b_scale,
       GetAllocationSliceForHlo(instr->operand(a_scale_index + 1)));
-#if GOOGLE_CUDA
   TF_ASSIGN_OR_RETURN(
       BufferAllocation::Slice c_scale,
       GetAllocationSliceForHlo(instr->operand(a_scale_index + 2)));
   TF_ASSIGN_OR_RETURN(
       BufferAllocation::Slice d_scale,
       GetAllocationSliceForHlo(instr->operand(a_scale_index + 3)));
-#else  // TENSORFLOW_USE_ROCM
-  BufferAllocation::Slice c_scale;
-  BufferAllocation::Slice d_scale;
-#endif
+  BufferAllocation::Slice aux, bias, d_amax;
```

Comment:
Can we streamline this logic? E.g. the case when has_vector_bias is true is repeated 4 times.

Reply:
Fixed.